research/conditional/moe_layers/expert_choice.py

```diff
@@ -86,6 +88,15 @@ def __init__(
     def forward(self, x: torch.Tensor):
         # x is (batch, seq_len, dmodel)
         batch_size, seq_len = x.shape[0], x.shape[1]
+        orig_bs, orig_seq_len = batch_size, seq_len
+
+        if self.simulate_groups > 1:
+            assert batch_size % self.simulate_groups == 0
+            batch_size, seq_len = (
+                batch_size // self.simulate_groups,
+                seq_len * self.simulate_groups,
+            )
+            x = x.reshape(batch_size, seq_len, self.dmodel)
```

Comment:
Nice hack. I think we can keep it for now, but in the future I would prefer a separate dimension for `group_size`. Maybe we can add `TODO` to change it in the future.

Reply:
the thing is, that separate dimension would require rewriting a lot of code, and it wont easily translate to different implementations, for example without einsum. Using separate dimension was my first attempt.